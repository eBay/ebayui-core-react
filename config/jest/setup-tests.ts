import '@testing-library/jest-dom'

declare var global: typeof globalThis;

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: () => {} });
}
