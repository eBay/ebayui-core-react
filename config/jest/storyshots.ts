import { render } from "@testing-library/react";
import initStoryshotsAddon from "@storybook/addon-storyshots";
import type { StoryshotsOptions } from "@storybook/addon-storyshots/dist/ts3.9/api/StoryshotsOptions";

const reactTestingLibrarySerializer = {
    print: (val, serialize, indent) => serialize(val.asFragment()),
    test: (val) => val && val.hasOwnProperty("container"),
};

export function initStoryshots(config: StoryshotsOptions) {
    initStoryshotsAddon({
        ...config,
        renderer: render,
        snapshotSerializers: [reactTestingLibrarySerializer],
    });
}
