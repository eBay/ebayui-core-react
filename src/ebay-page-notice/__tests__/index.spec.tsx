import React from 'react';
import { render } from '@testing-library/react'
import { initStoryshots } from '../../../config/jest/storyshots';
import { EbayNoticeContent, EbayPageNotice, EbayPageNoticeTitle } from '../';
import userEvent from '@testing-library/user-event'

jest.mock('../../common/random-id', () => ({ randomId: () => 'abc123' }))

describe('<EbayPageNotice>', () => {
    describe('when a page title exists', () => {
        it('should render the title within the notice content', () => {
            const noticeContentSelector = 'div.page-notice__main'
            const wrapper = render(
                <EbayPageNotice status="confirmation">
                    <EbayNoticeContent>
                        <EbayPageNoticeTitle>Here is the title</EbayPageNoticeTitle>
                        <p id="details">Here is some details</p>
                    </EbayNoticeContent>
                </EbayPageNotice>
            )
            expect(wrapper.container.querySelector(noticeContentSelector)).toBeTruthy()
            expect(wrapper.container.querySelector(`${noticeContentSelector} > p#details`)).toBeTruthy()
            expect(wrapper.getByText('Here is the title')).toBeTruthy()
        })
    })

    describe('when aria-label is given', () => {
        it('should add the text to the icon as an aria-label', () => {
            const wrapper = render(
                <EbayPageNotice status="confirmation" aria-label="Success Confirmation">
                    <EbayNoticeContent>
                        <p>You did it.</p>
                    </EbayNoticeContent>
                </EbayPageNotice>
            )
            expect(wrapper.container.querySelector('svg')).toHaveAttribute('role', 'img')
        })
    })

    describe('on using the notice with no content', () => {
        it('should throw an error', () => {
            jest.spyOn(console, 'error').mockImplementation(() => null)
            expect(() => {
                render(<EbayPageNotice />)
            }).toThrowError(`EbayPageNotice: Please use a EbayNoticeContent that defines the content of the notice`);
            (console.error as jest.Mock).mockRestore()
        })
    })

    describe('when allyDismissText is provided...', () => {
        let wrapper;
        let dismissButton;
        const dismissMock = jest.fn();

        beforeEach( async () => {
            wrapper = render(
                <EbayPageNotice status="information" aria-label="Information" a11yDismissText="Close" onDismiss={dismissMock}>
                    <EbayNoticeContent>
                        <p>Content</p>
                    </EbayNoticeContent>
                </EbayPageNotice>
            )
            dismissButton = await wrapper.getByRole('button', {name:'Close'});
        })

        it( 'should add a close button with the provided label.', () => {
            expect( dismissButton ).not.toBeNull();
        })

        it( 'should hide the notice when the dismiss button is clicked.', async () => {
            expect( wrapper.getByRole('region': {name:'Information'}) ).toBeVisible();
            await dismissButton.click();
            expect( wrapper.queryByRole('region': {name:'Information'}) ).toBeNull();
            expect( dismissMock ).toHaveBeenCalled();
        })

        it( 'should hide the notice when the user focuses the dismiss button and presses space', async () => {
            expect( wrapper.getByRole('region': {name:'Information'}) ).toBeVisible();
            await dismissButton.focus();
            userEvent.type( dismissButton, " " );
            expect( wrapper.queryByRole('region': {name:'Information'}) ).toBeNull();
            expect( dismissMock ).toHaveBeenCalled();
        })
    })
})

initStoryshots({
    config: ({ configure }) =>
        configure(() => {
            require('./index.stories')
        }, module)
})
