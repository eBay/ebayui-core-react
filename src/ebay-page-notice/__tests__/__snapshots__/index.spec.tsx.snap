// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots ebay-page-notice Attention message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="attention-status"
    class="page-notice page-notice--attention"
    role="region"
  >
    <div
      class="page-notice__header"
      id="attention-status"
    >
      <svg
        aria-label="Attention"
        class="icon icon--attention-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-attention-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Error.
      </h2>
      <p>
        Please take another look at the following:
        <br />
        <a
          href="#"
        >
          Card number
        </a>
        , 
        <a
          href="#"
        >
          Expiration date
        </a>
         & 
        <a
          href="#"
        >
          Security code
        </a>
        .
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Confirmation message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="confirmation-status"
    class="page-notice page-notice--confirmation"
    role="region"
  >
    <div
      class="page-notice__header"
      id="confirmation-status"
    >
      <svg
        aria-label="Success"
        class="icon icon--confirmation-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-confirmation-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Congrats!
      </h2>
      <p>
        You just listed 
        <a
          href="#"
        >
          Spam and Eggs From the Cows Perspective
        </a>
         (paperback).
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Dismissible message with CTA 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="information-status"
    class="page-notice page-notice--information"
    role="region"
  >
    <div
      class="page-notice__header"
      id="information-status"
    >
      <svg
        aria-label="Congratulations"
        class="icon icon--information-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-information-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Your order's in!
      </h2>
      <p>
        We'll email updates to jonsnow@gmail.com. You should get it by Thu, Sept 22.
      </p>
    </div>
    <p
      class="page-notice__cta"
    >
      <a
        href="https://ebay.com"
      >
        Action
      </a>
    </p>
    <div
      class="page-notice__footer"
    >
      <button
        aria-label="Close"
        class="fake-link page-notice__dismiss"
      >
        <svg
          aria-hidden="true"
          class="icon icon--close-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-close-small"
          />
        </svg>
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Dismissible notice 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="information-status"
    class="page-notice page-notice--information"
    role="region"
  >
    <div
      class="page-notice__header"
      id="information-status"
    >
      <svg
        aria-label="Information"
        class="icon icon--information-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-information-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Good news!
      </h2>
      <p>
        You get free shipping on your next pair of shoes! 
        <a
          href="#"
        >
          Learn more
        </a>
        .
      </p>
    </div>
    <div
      class="page-notice__footer"
    >
      <button
        aria-label="Close"
        class="fake-link page-notice__dismiss"
      >
        <svg
          aria-hidden="true"
          class="icon icon--close-small"
          focusable="false"
          height="16px"
          width="16px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="#icon-close-small"
          />
        </svg>
      </button>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Information message 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="information-status"
    class="page-notice page-notice--information"
    role="region"
  >
    <div
      class="page-notice__header"
      id="information-status"
    >
      <svg
        aria-label="Information"
        class="icon icon--information-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-information-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Good news!
      </h2>
      <p>
        You get free shipping on your next pair of shoes! 
        <a
          href="#"
        >
          Learn more
        </a>
        .
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Message with footer 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="confirmation-status"
    class="page-notice page-notice--confirmation"
    role="region"
  >
    <div
      class="page-notice__header"
      id="confirmation-status"
    >
      <svg
        aria-label="Congratulations"
        class="icon icon--confirmation-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-confirmation-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      <h2
        class="page-notice__title"
      >
        Your order's in!
      </h2>
      <p>
        We'll email updates to jonsnow@gmail.com. You should get it by Thu, Sept 22.
      </p>
    </div>
    <div
      class="page-notice__footer"
    >
      <a
        href="https://ebay.com"
      >
        Action
      </a>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Simple usage 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="general-status"
    class="page-notice "
    role="region"
  >
    <div
      class="page-notice__main"
    >
      text message
    </div>
  </section>
</DocumentFragment>
`;

exports[`Storyshots ebay-page-notice Simple usage with id 1`] = `
<DocumentFragment>
  <section
    aria-labelledby="main-page-notice"
    class="page-notice page-notice--confirmation"
    role="region"
  >
    <div
      class="page-notice__header"
      id="main-page-notice"
    >
      <svg
        aria-label="Success"
        class="icon icon--confirmation-filled-small"
        focusable="false"
        height="16px"
        role="img"
        width="16px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <use
          xlink:href="#icon-confirmation-filled-small"
        />
      </svg>
    </div>
    <div
      class="page-notice__main"
    >
      text message
    </div>
  </section>
</DocumentFragment>
`;
